art/defences/trainer/adversarial_trainer_oaat_pytorch.py

```diff
@@ -0,0 +1,1224 @@
+from art.estimators.classification.pytorch import PyTorchClassifier
+from art.data_generators import DataGenerator
+from art.attacks.attack import EvasionAttack
+from art.utils import check_and_transform_label_format, random_sphere
+from art.config import ART_NUMPY_DTYPE
+
+if TYPE_CHECKING:
+    import torch
+
+logger = logging.getLogger(__name__)
+EPS = 1e-8  # small value required for avoiding division by zero and for KLDivLoss to make probability vector non-zero
+
+
+class AdversarialTrainerOAATPyTorch(AdversarialTrainerOAAT):
+    """
+    Class performing adversarial training following Oracle Aligned Adversarial Training (OAAT) protocol.
+
+    | Paper link: https://link.springer.com/chapter/10.1007/978-3-031-20065-6_18
+    """
+
+    def __init__(
+        self,
+        classifier: PyTorchClassifier,
+        proxy_classifier: PyTorchClassifier,
+        lpips_classifier: PyTorchClassifier,
+        list_avg_models: List[PyTorchClassifier],
+        attack: EvasionAttack,
+        train_params: dict,
+    ):
+        """
+        Create an :class:`.AdversarialTrainerOAATPyTorch` instance.
+
+        :param classifier: Model to train adversarially.
+        :param proxy_classifier: Model for adversarial weight perturbation.
+        :param lpips_classifier: Weight averaging model for calculating activations.
+        :param list_avg_models: list of models for weight averaging.
+        :param attack: attack to use for data augmentation in adversarial training.
+        :param train_params: training parmaters' dictionary related to adversarial training
+        """
+        super().__init__(classifier, proxy_classifier, lpips_classifier, list_avg_models, attack, train_params)
+        self._classifier: PyTorchClassifier
+        self._proxy_classifier: PyTorchClassifier
+        self._lpips_classifier: PyTorchClassifier
+        self._list_avg_models: List[PyTorchClassifier]
+        self._attack: EvasionAttack
+        self._train_params: dict
+        self._apply_wp: bool
+        self._apply_lpips_pert: bool
+
+    def fit(
+        self,
+        x: np.ndarray,
+        y: np.ndarray,
+        validation_data: Optional[Tuple[np.ndarray, np.ndarray]] = None,
+        batch_size: int = 128,
+        nb_epochs: int = 20,
+        **kwargs,
+    ):  # pylint: disable=W0221
+        """
+        Train a model adversarially with OAAT protocol.
+        See class documentation for more information on the exact procedure.
+
+        :param x: Training set.
+        :param y: Labels for the training set.
+        :param validation_data: Tuple consisting of validation data, (x_val, y_val)
+        :param batch_size: Size of batches.
+        :param nb_epochs: Number of epochs to use for trainings.
+        :param kwargs: Dictionary of framework-specific arguments. These will be passed as such to the `fit` function of
+                                  the target classifier.
+        """
+        import os
```

Comment:
Please move the import of 'os' to the top-level.

Reply:
moved